trie-db/src/iterator.rs

```diff
@@ -0,0 +1,647 @@
+                    },
+                    Node::NibbledBranch(slice, nodes, _) => {
+                        if !partial.starts_with(&slice) {
+                            if slice < partial {
+                                crumb.status = Status::Exiting;
+                                self.key_nibbles.append_partial(slice.right());
+                                self.key_nibbles.push((nibble_ops::NIBBLE_LENGTH - 1) as u8);
+                            }
+                            return Ok(());
+                        }
+
+                        full_key_nibbles += slice.len();
+                        partial = partial.mid(slice.len());
+
+                        if partial.is_empty() {
+                            return Ok(())
+                        }
+
+                        let i = partial.at(0);
+                        crumb.status = Status::AtChild(i as usize);
+                        self.key_nibbles.append_partial(slice.right());
+                        self.key_nibbles.push(i);
+
+                        if let Some(child) = nodes[i as usize] {
+                            full_key_nibbles += 1;
+                            partial = partial.mid(1);
+
+                            let prefix = key.back(full_key_nibbles);
+                            self.db.get_raw_or_lookup(child, prefix.left())?
+                        } else {
+                            return Ok(())
+                        }
+                    },
+                    Node::Empty => {
+                        if !partial.is_empty() {
+                            crumb.status = Status::Exiting;
+                        }
+                        return Ok(())
+                    },
+                }
+            };
+
+            node_data = next_node_data;
+            inline = next_inline;
+        }
+    }
+
+    /// Descend into a payload.
+    fn descend<'b, 'c>(&'b mut self, node_data: &'c [u8], inline: bool)
+        -> Result<Node<'c>, TrieHash<L>, CError<L>>
+    {
+        let node_hash = if inline {
+            None
+        } else {
+            Some(L::Hash::hash(node_data))
+        };
+        let node = L::Codec::decode(&node_data)
+            .map_err(|e| Box::new(TrieError::DecoderError(node_hash.unwrap_or_default(), e)))?;
+        self.trail.push(Crumb {
+            hash: node_hash,
+            status: Status::Entering,
+            node: Rc::new(node.clone().into()),
+        });
+        Ok(node)
+    }
+}
+
+impl<'a, L: TrieLayout> TrieIterator<L> for TrieDBNodeIterator<'a, L> {
+    fn seek(&mut self, key: &[u8]) -> Result<(), TrieHash<L>, CError<L>> {
+        self.trail.clear();
+        self.key_nibbles.clear();
+        let root_node = self.db.root_data()?;
+        self.seek(root_node, false, NibbleSlice::new(key.as_ref()))
+    }
+}
+
+impl<'a, L: TrieLayout> Iterator for TrieDBNodeIterator<'a, L> {
+    type Item = Result<(NibbleVec, Option<TrieHash<L>>, Rc<OwnedNode>), TrieHash<L>, CError<L>>;
```

Comment:
I do not really understand the need for the node hash (relation between nodes could be done through the nibblevec), simply knowing if inline seems ok.
Returning the NibbleVec involves a clone that is not needed in most case (considering the standard iterator), this does not feels super satisfactory but the lifetime involve seems rather tricky (I think there is a reason why an rc is use for the node).

Reply:
See https://github.com/paritytech/trie/pull/33#discussion_r340027862 for discussion of the hash. Yes, the key prefix needs to be owned due to lifetime issues and that is also why I chose to return an `Rc` instead of doing a clone on the `OwnedNode`.